### PR TITLE
Expand the new block menu to make it easier for beginners and more visual

### PR DESCRIPTION
**Pull request by aickin**
Modified the New Block menu to make it more visual and friendlier to Dust newcomers. Current menu looks like this:

image

and after this PR the menu looks like this:

image

I did this because I thought that the menu that just showed the block type identifiers was probably not very friendly to folks who are just starting out with Dust and maybe don't know what it's capabilities are. I thought that a more human-readable name paired with a functional description might make make Dust's value proposition clearer.

After adding this menu and playing with it for a bit, I decided that I should probably also replace the block type tokens on the blocks themselves, so I replaced them with the icons. On mouseover, the icons show the human readable block type ("JavaScript", "Large Language Model (LLM)", etc):

image

I'm not totally certain I got the descriptions right, particularly on Data and Map Reduce blocks; please feel free to suggest edits. Also totally happy to hear feedback on the icon choice.

Note that this PR would need to change slightly if and when Replit blocks (PR #40) are added. It will not be a big change, however.

### CODE REVIEW

**Review by spolu — CHANGES_REQUESTED** [review]
Adding the description makes a ton of sense. I think it's a bit too early for the icons.

How would you feel about keeping the original block names in the menu, removing the icons but keeping the text as a starter?

**Comment by aickin** [comment]
When you say "too early" for icons, what does you mean? Is there a concern that we wouldn't be able to pull them out later if we want?

> How would you feel about keeping the original block names in the menu, removing the icons but keeping the text as a starter?

When you say "original block names", do you mean the names like "llm", "code", "search"? I certainly can put those back, but I feel like they are pretty unfriendly to folks who aren't familiar with the Dust codebase. Is there a reason for using the short lower case names rather than human-readable ones?

**Comment by spolu** [comment]
> When you say "too early" for icons, what does you mean? Is there a concern that we wouldn't be able to pull them out later if we want?

I think they're a bit too graphical and I'd rather stay on the side of keeping the UI as simple as possible for now.

> When you say "original block names", do you mean the names like "llm", "code", "search"? I certainly can put those back, but I feel like they are pretty unfriendly to folks who aren't familiar with the Dust codebase. Is there a reason for using the short lower case names rather than human-readable ones?

Yes remaining aligned with the underlying specification language.

With the added description I think it becomes much more friendly! :+1:

**Comment by aickin** [comment]
> keeping the UI as simple as possible for now.

I hear you, but I'd argue that for a lot of folks, adding icons does make UI simpler, as it gives them something to latch on to in a way that doesn't involve reading. I'll go ahead and take them out, because you are clear that they don't fit your vision, but maybe at some point in the future we can show them to some users and see how they react?

Concerning the names, how would you feel about using the human readable name, with the spec language name in brackets or parens, like: 

Large Language Model (llm)
or
JavaScript <code>

I know I'm pushing on this a bit, but I just feel like people using the Dust UI aren't going to know anything about the specification language. There isn't a way for them to access it right now and I don't think it's documented (and even if it were, we wouldn't want people to read it before using the UI), so the spec block names are not going to be particularly useful imo.

> With the added description I think it becomes much more friendly! 👍

Glad to hear it! 😃

**Comment by spolu** [comment]
Proposal: keep the original visual in place of the icons and keep the human readable title + description ?

**Comment by aickin** [comment]
Just trying to make sure I understand you completely. When you say "original visual", is there something else about the visuals of the original menu that you want to preserve (other than it not having icons)? Or do you mean basically take what I have in this PR with the human readable title and description and just get rid of the icon column? Happy to do either!

**Comment by spolu** [comment]
I think the map/reduce description needs a bit of work other ones look pretty good.

```
Map Reduce: Map over an array and execute a sequence of blocks in parallel.
```

maybe?

**Comment by aickin** [comment]
> Map Reduce: Map over an array and execute a sequence of blocks in parallel.

Sounds good. I literally don't know what those blocks are for; hadn't used them in a Dust app yet! 🤣

**Comment by spolu** [comment]
> Just trying to make sure I understand you completely. When you say "original visual", is there something else about the visuals of the original menu that you want to preserve (other than it not having icons)? Or do you mean basically take what I have in this PR with the human readable title and description and just get rid of the icon column? Happy to do either!

I think I value having the same visual in the block and the menu (which you replaced by icons).

**Comment by aickin** [comment]
Ok, I think I understand what you mean. I'll work on that now!

**Comment by spolu** [comment]
\o/

**Comment by aickin** [comment]
Ok, pushed some commits, and now the menu now looks like this: 
image
And the block UI is back to what it was before the PR. Is this what you were looking for?

File: front/components/app/NewBlock.jsx [hunk 1]
```diff
@@ -43,44 +85,41 @@ export default function NewBlock({ spec, disabled, onClick, direction }) {
...
       >
-        <Menu.Items
+        <Popover.Panel
+          ref={setPopperElement}
+          style={styles.popper}
+          {...attributes.popper}
           className={classNames(
-            "absolute shadow left-1 z-10 mt-1 w-32 origin-top-right rounded-md bg-white ring-1 ring-black ring-opacity-5 focus:outline-none",
-            direction == "up" ? "bottom-10" : ""
+            "absolute shadow left-1 z-10 mt-1  origin-top-right rounded-md bg-white ring-1 ring-black ring-opacity-5 focus:outline-none",
```
Comment (spolu) [inline front/components/app/NewBlock.jsx:93]: double space

File: front/components/app/NewBlock.jsx [hunk 2]
```diff
@@ -1,39 +1,81 @@
...
-    { type: "llm", display: ["llm"] },
-    { type: "code", display: ["code"] },
-    { type: "search", display: ["search"] },
-    { type: "map_reduce", display: ["map", "reduce"] },
+    {
+      type: "llm",
+      typeNames: ["llm"],
+      name: "Large Language Model (LLM)",
+      description:
+        "Ask an artificial intelligence Large Language Model to complete a prompt for you. Current LLMs supported: OpenAI's GPT-3 and Cohere.",
```
Comment (spolu) [inline front/components/app/NewBlock.jsx:22]: Query a Large Language Model to complete a prompt for you.

File: front/components/app/NewBlock.jsx [hunk 3]
```diff
@@ -1,39 +1,81 @@
...
+    },
   ];
   if (!containsInput) {
-    blocks.splice(0, 0, { type: "input", display: ["input"] });
+    blocks.splice(0, 0, {
+      type: "input",
+      typeNames: ["input"],
+      name: "Input",
+      description:
+        "Select a dataset that is used as the input to this Dust app, like the arguments to a function. Each element in the dataset kicks off a separate parallel run of the Dust app.",
```
Comment (spolu) [inline front/components/app/NewBlock.jsx:59]: Select a dataset of inputs used for the design your Dust app. Each element in the dataset kicks off a separate parallel execution of the Dust app.

File: front/components/app/NewBlock.jsx [hunk 4]
```diff
@@ -1,39 +1,81 @@
...
+      name: "Large Language Model (LLM)",
+      description:
+        "Ask an artificial intelligence Large Language Model to complete a prompt for you. Current LLMs supported: OpenAI's GPT-3 and Cohere.",
+    },
+    {
+      type: "data",
+      typeNames: ["data"],
+      name: "Data",
+      description:
+        "Load a dataset to be used for every run of the Dust app. Typically used to seed a few-shot prompt to an LLM block.",
```
Comment (spolu) [inline front/components/app/NewBlock.jsx:29]: Load a dataset and outputs its element as an array. Typically used to seed a few-shot prompt to an LLM block.

**Review by spolu — CHANGES_REQUESTED** [review]
Left a few comments on the description :+1:

File: front/components/app/NewBlock.jsx (hunk 1 above)
Comment (aickin) [reply front/components/app/NewBlock.jsx]: Fix in ba034f3.

File: front/components/app/NewBlock.jsx (hunk 2 above)
Comment (aickin) [reply front/components/app/NewBlock.jsx]: Fix in ba034f3.

File: front/components/app/NewBlock.jsx (hunk 4 above)
Comment (aickin) [reply front/components/app/NewBlock.jsx]: Fix in ba034f3.

**Pull request merged**